CHAID/stats.py

```diff
@@ -73,40 +77,103 @@ def best_cat_split(self, ind, dep):
                     for dep_v in all_dep:
                         freq[col][dep_v] = dep.weights[(ind_var.arr == col) * (dep.arr == dep_v)].sum()
 
+
+            # merge all most similar, too small categories together until have sufficient base size
+            # take the two smallest, if both smaller than min child node size, merge
+            break_out = False
+            has_merged = 1
+            while has_merged == 1:
```

Comment:
This kind of flagging seems super gross. Perhaps switching the last if to a negative and doing a break would be better. Also there are so many booleans flying round, some well named variables would be 👌 
```python
while True:
  data = np.array([ [k, sum(v.values())] for k,v in freq.items() ])
  data_sorted = data[data[:,1].argsort()]

  too_few_respondents =  sum(data_sorted[:, 1]) < self.min_child_node_size
  nodes_below_min = data_sorted[0:2, 1] < self.min_child_node_size
  no_splits_available = len(data_sorted[:, 1]) == 2 and nodes_below_min.any()

  if too_few_respondents or no_splits_available:
    split.invalid_reason = InvalidSplitReason.MIN_CHILD_NODE_SIZE
    break

  if not nodes_below_min.all():
    break

  first  = min(data_sorted[0, 0], data_sorted[1, 0])
  # ...
```

Reply:
We'll also want to hit each one of those booleans separately in its own test case.